File: godtools-sync/src/main/java/org/cru/godtools/sync/service/FollowupService.java
package org.cru.godtools.sync.service;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

import org.ccci.gto.android.common.util.NumberUtils;
import org.cru.godtools.base.model.Event;
import org.cru.godtools.model.Followup;
import org.cru.godtools.sync.GodToolsSyncService;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.keynote.godtools.android.db.GodToolsDao;

import static org.cru.godtools.base.model.Event.Id.FOLLOWUP_EVENT;

public final class FollowupService {
    private static final String FIELD_NAME = "name";
    private static final String FIELD_EMAIL = "email";
    private static final String FIELD_DESTINATION = "destination_id";

    private final Context mContext;
    private final GodToolsDao mDao;

    private FollowupService(@NonNull final Context context) {
        mContext = context;
        mDao = GodToolsDao.getInstance(context);
        EventBus.getDefault().register(this);
    }

    @Nullable
    @SuppressLint("StaticFieldLeak")
    private static FollowupService sInstance;
    @NonNull
    @MainThread
    public static FollowupService start(@NonNull final Context context) {
        synchronized (FollowupService.class) {
            if (sInstance == null) {
                sInstance = new FollowupService(context.getApplicationContext());
            }
        }

        return sInstance;
    }

    @WorkerThread
    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    public void onContentEvent(@NonNull final Event event) {
        if (FOLLOWUP_EVENT.equals(event.id)) {
            final Followup followup = new Followup();
            followup.setName(event.fields.get(FIELD_NAME));
            followup.setEmail(event.fields.get(FIELD_EMAIL));
            followup.setLanguageCode(event.locale);
            followup.setDestination(NumberUtils.toLong(event.fields.get(FIELD_DESTINATION), null));
            mDao.insertNew(followup);

            // trigger a sync of followups
            GodToolsSyncService.syncFollowups(mContext).run();
        }
    }
}

File: ui/base-tool/src/main/kotlin/org/cru/godtools/base/tool/service/FollowupService.kt
package org.cru.godtools.base.tool.service

import dagger.Lazy
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.ccci.gto.android.common.dagger.getValue
import org.cru.godtools.base.tool.model.Event
import org.cru.godtools.db.repository.FollowupsRepository
import org.cru.godtools.model.Followup
import org.cru.godtools.shared.renderer.state.State
import org.cru.godtools.shared.tool.parser.model.EventId
import org.cru.godtools.sync.GodToolsSyncService
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

private const val FIELD_NAME = "name"
private const val FIELD_EMAIL = "email"
private const val FIELD_DESTINATION = "destination_id"

@Singleton
class FollowupService @Inject internal constructor(
    eventBus: EventBus,
    followupsRepository: Lazy<FollowupsRepository>,
    syncService: Lazy<GodToolsSyncService>
) {
    private val coroutineScope = CoroutineScope(Dispatchers.Default)

    private val followupsRepository by followupsRepository
    private val syncService by syncService

    @Subscribe
    fun onContentEvent(event: Event) {
        if (event.id != EventId.FOLLOWUP) return

        coroutineScope.launch {
            val followup = Followup(
                destination = event.fields[FIELD_DESTINATION]?.toLongOrNull() ?: return@launch,
                languageCode = event.locale ?: return@launch,
                name = event.fields[FIELD_NAME],
                email = event.fields[FIELD_EMAIL] ?: return@launch,
            )

            // only store this followup if it's valid
            if (followup.isValid) {
                followupsRepository.createFollowup(followup)
                @Suppress("DeferredResultUnused")
                syncService.syncFollowupsAsync()
            }
        }
    }

    fun handleSubmitFormEvent(data: State.Event.SubmitForm, locale: Locale?) {
        coroutineScope.launch {
            val followup = Followup(
                destination = data.fields[FIELD_DESTINATION]?.toLongOrNull() ?: return@launch,
                languageCode = locale ?: return@launch,
                name = data.fields[FIELD_NAME],
                email = data.fields[FIELD_EMAIL] ?: return@launch,
            )

            // only store this followup if it's valid
            if (followup.isValid) {
                followupsRepository.createFollowup(followup)
                @Suppress("DeferredResultUnused")
                syncService.syncFollowupsAsync()
            }
        }
    }

    init {
        eventBus.register(this)
        coroutineScope.launch {
            @Suppress("DeferredResultUnused")
            this@FollowupService.syncService.syncFollowupsAsync()
        }
    }
}
